vanetza/security/naive_certificate_provider.cpp

```diff
@@ -0,0 +1,142 @@
+            break;
+        }
+    }
+
+    return m_own_certificate;
+}
+
+const ecdsa256::PrivateKey& NaiveCertificateProvider::own_private_key()
+{
+    return m_own_key_pair.private_key;
+}
+
+const Certificate& NaiveCertificateProvider::root_certificate()
+{
+    return m_root_certificate;
+}
+
+Certificate NaiveCertificateProvider::generate_authorization_ticket()
+{
+    // create certificate
+    Certificate certificate;
+
+    // section 6.1 in TS 103 097 v1.2.1
+    certificate.signer_info = m_root_certificate_hash;
+
+    // section 6.3 in TS 103 097 v1.2.1
+    certificate.subject_info.subject_type = SubjectType::Authorization_Ticket;
+    // section 7.4.2 in TS 103 097 v1.2.1, subject_name implicit empty
+
+    // set assurance level
+    certificate.subject_attributes.push_back(SubjectAssurance(0x00));
+
+    // section 7.4.1 in TS 103 097 v1.2.1
+    // set subject attributes
+    // set the verification_key
+    Uncompressed coordinates;
+    coordinates.x.assign(m_own_key_pair.public_key.x.begin(), m_own_key_pair.public_key.x.end());
+    coordinates.y.assign(m_own_key_pair.public_key.y.begin(), m_own_key_pair.public_key.y.end());
+    EccPoint ecc_point = coordinates;
+    ecdsa_nistp256_with_sha256 ecdsa;
+    ecdsa.public_key = ecc_point;
+    VerificationKey verification_key;
+    verification_key.key = ecdsa;
+    certificate.subject_attributes.push_back(verification_key);
+
+    // section 6.7 in TS 103 097 v1.2.1
+    // set validity restriction
+    StartAndEndValidity start_and_end;
+    start_and_end.start_validity = convert_time32(m_time_now - std::chrono::hours(1));
+    start_and_end.end_validity = convert_time32(m_time_now + std::chrono::hours(23));
+    certificate.validity_restriction.push_back(start_and_end);
+
+    // set signature
+    ByteBuffer data_buffer = convert_for_signing(certificate);
+    certificate.signature = m_crypto_backend.sign_data(m_root_key_pair.private_key, data_buffer);
+
+    return certificate;
+}
+
+Certificate NaiveCertificateProvider::generate_root_certificate()
+{
+    // create certificate
+    Certificate certificate;
+
+    // section 6.1 in TS 103 097 v1.2.1
+    certificate.signer_info = nullptr; /* self */
+
+    // section 6.3 in TS 103 097 v1.2.1
+    certificate.subject_info.subject_type = SubjectType::Root_Ca;
+
+    // section 7.4.2 in TS 103 097 v1.2.1
+    std::vector<unsigned char> subject(m_root_certificate_subject.c_str(), m_root_certificate_subject.c_str() + m_root_certificate_subject.size() + 1);
+    certificate.subject_info.subject_name = subject;
+
+    // section 6.6 in TS 103 097 v1.2.1 - levels currently undefined
+    certificate.subject_attributes.push_back(SubjectAssurance(0x00));
+
+    // section 7.4.1 in TS 103 097 v1.2.1
+    // set subject attributes
+    // set the verification_key
+    Uncompressed coordinates;
+    coordinates.x.assign(m_root_key_pair.public_key.x.begin(), m_root_key_pair.public_key.x.end());
```

Comment:
@riebl This currently segfaults in several unit tests, but I fail to see why for quite some time now. Could you have a look?

```
#11 0x00005555559b05f4 in std::vector<unsigned char, std::allocator<unsigned char> >::assign<unsigned char const*, void> (this=0x7fffffffd4d0, 
    __first=0x20 <error: Cannot access memory at address 0x20>, 
    __last=0x40 <error: Cannot access memory at address 0x40>)
    at /usr/include/c++/7/bits/stl_vector.h:521

#12 0x00005555559afbeb in vanetza::security::NaiveCertificateProvider::generate_root_certificate (this=0x555555d4c0d0)
    at /home/kelunik/GitHub/riebl/vanetza/vanetza/security/naive_certificate_provider.cpp:112
```

If I try to avoid the reference for the root pair and use a freshly generated pair like for the `own_key_pair`, then it fails with that error instead:

```
C++ exception with description "CryptoMaterial: this object contains invalid values" thrown in the test fixture's constructor.
```

Reply:
I think I have found the reason for your segfaults, see my branch kelunik-basic-cert-mngr at 8709f1ed982f53815abbf0d47e7fd4f9f09ed83e.
However, some unit tests still fail (no segfault).

Reply:
Thanks, I've just pushed a commit that fixes those failures as well. It just used the wrong certificate as signer.